servatrice/src/serversocketinterface.cpp

```diff
@@ -1000,3 +1002,55 @@ Response::ResponseCode ServerSocketInterface::cmdReloadConfig(const Command_Relo
     settingsCache->sync();
     return Response::RespOk;
 }
+
+Response::ResponseCode ServerSocketInterface::cmdAdjustMod(const Command_AdjustMod &cmd, ResponseContainer & /*rc*/) {
+
+    QString adminName = QString::fromStdString(cmd.admin_name());
+    QString userName = QString::fromStdString(cmd.user_name());
+
+    if (cmd.should_be_mod()) {
+        logDebugMessage("Received admin command: promote user");
+        QSqlQuery *query = sqlInterface->prepareQuery(
+                "update {prefix}_users set admin = :adminlevel where name = :username");
+        query->bindValue(":adminlevel", 2);
+        query->bindValue(":username", userName);
+        if (!sqlInterface->execSqlQuery(query))
+            return Response::RespInternalError;
+
```

Comment:
Don't make assumptions about what the client can do, someone can always
modify it. Code the server and protocol for the worst case.

Under normal operation I could perform the promote and before the server
gets my message the target could disconnect.so it should seamlessly handle
promoting offline users. Similarly we should be able to demote bad mods
when offline

On Thu, Aug 20, 2015, 9:57 PM woogerboy21 <notifications@github.com> wrote:

> In servatrice/src/serversocketinterface.cpp
> <https://github.com/Cockatrice/Cockatrice/pull/1391#discussion_r37600432>:
>
> > @@ -1000,3 +1002,55 @@ Response::ResponseCode ServerSocketInterface::cmdReloadConfig(const Command_Relo
> >      settingsCache->sync();
> >      return Response::RespOk;
> >  }
> > +
> > +Response::ResponseCode ServerSocketInterface::cmdAdjustMod(const Command_AdjustMod &cmd, ResponseContainer & /*rc*/) {
> > +
> > +    QString adminName = QString::fromStdString(cmd.admin_name());
> > +    QString userName = QString::fromStdString(cmd.user_name());
> > +
> > +    if (cmd.should_be_mod()) {
> > +        logDebugMessage("Received admin command: promote user");
> > +        QSqlQuery *query = sqlInterface->prepareQuery(
> > +                "update {prefix}_users set admin = :adminlevel where name = :username");
> > +        query->bindValue(":adminlevel", 2);
> > +        query->bindValue(":username", userName);
> > +        if (!sqlInterface->execSqlQuery(query))
> > +            return Response::RespInternalError;
> > +
>
> The command is sent from a menu that shows a clients name listed. The only
> way the command could be fired from a menu with an offline client is from
> your buddy list. Which will show the user as grey but still be able to pull
> the name from the list.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Cockatrice/Cockatrice/pull/1391/files#r37600432>.
>